File: Filtering/ANF/DigitalNF.m
%% Digital Sequential Notch Filter implementation for ASTRAv2.
%  All IMU sensors are meant to be passed to this array always. Includes a
%  GND switch option so the filter does not run on the ground when no
%  vibrations are present. 

function OUT = DigitalNF(IN, GND, THRUST, dT)    

% Initialize past output / input memory column 
persistent X1 X2 Y1 Y2
if isempty(X1)
    X1 = ones(size(IN, 1), 2) .* IN;
    X2 = X1;
    Y1 = X1;
    Y2 = X1;
end

% Center Freq. vs Thrust Tracks, Sampling Freq, and Notch Width
TRACK = [1.3525    42.6278;
         2.6867    84.8000];
fs = 1 / dT;
width1 = 35;     % Hz

% Sequential Notch Filter
if ~GND
    %% Notch #1
        % Setup first notch at constant frequency
        f0 = TRACK(1, 1) * THRUST + TRACK(1, 2);    %Hz
        w0 = 2 * pi * f0 / fs;
        r = exp(-pi * width1 / fs);

        % Calculate Normalization Gain
        G_num = 1 - 2*r*cos(w0) + r^2;
        G_den = 2 - 2*cos(w0);
        G = G_num / G_den;

        % Filter Out
        NUM = G * (IN - 2*cos(w0)*X1(:, 1) + X1(:, 2));
        DEN = 2*r*cos(w0)*Y1(:,1) - r^2*Y1(:,2);
        OUT = NUM + DEN;
    
        % Update memory for first Notch
        X1 = [IN    X1(:, 1)];
        Y1 = [OUT   Y1(:, 1)];
        IN = OUT;
    
    %% 2nd Order Butterworth Filter
        % Set up a 2nd Order Butterworth Filter
        f0 = 20;
        C = tan(pi * f0 / fs);
        A1 = 2*(C^2 - 1) / (1 + sqrt(2)*C + C^2);
        A2 = (1 - sqrt(2)*C + C^2) / (1 + sqrt(2)*C + C^2);
        B0 = C^2 / (1 + sqrt(2)*C + C^2);
        B1 = 2 * B0;
        B2 = B0;
        OUT = B0*IN + B1*X2(:, 1) + B2*X2(:,2) - A1*Y2(:,1) - A2*Y2(:,2);

        % Update memory for second Notch
        X2 = [IN    X2(:, 1)];
        Y2 = [OUT   Y2(:, 1)];
else
    % The output is just the input (passthrough)
    OUT = IN; 
    % --- Update states sequentially for a smooth switch-on ---

    % Stage 1: History is updated with IN and OUT
    X1 = [IN    X1(:, 1)];
    Y1 = [OUT   Y1(:, 1)];

    % The input to stage 2 is the output of stage 1
    IN_2 = OUT;
    OUT_2 = IN_2; % Passthrough for stage 2

    % Stage 2: History is updated
    X2 = [IN_2  X2(:, 1)];
    Y2 = [OUT_2 Y2(:, 1)];
    OUT = OUT_2;
end

File: cpp_mex/DigitalNF.cpp
// Digital Sequential Notch Filter implementation for ASTRAv2.
// All IMU sensors are meant to be passed to this array always. Includes a
// GND switch option so the filter does not run on the ground when no
// vibrations are present.

#include "math.h"
#include "matlab_funcs.h"

Vector9 FilterStage(Vector9 IN, float f0, float fs, float width, int stage, Matrix9_4 &X, Matrix9_4 &Y) {
  float w0 = 2 * M_PI * f0 / fs;
  float r = exp(-M_PI * width / fs);

  // Calculate Normalization Gain
  float G_num = 1 - 2 * r * cos(w0) + r * r;
  float G_den = 2 - 2 * cos(w0);
  float G = G_num / G_den;

  // Filter Out
  Vector9 NUM = G * (IN - 2 * cos(w0) * X.block<9, 1>(0, 2 * stage) + X.block<9, 1>(0, 2 * stage + 1));
  Vector9 DEN = 2 * r * cos(w0) * Y.block<9, 1>(0, 2 * stage) - r * r * Y.block<9, 1>(0, 2 * stage + 1);
  Vector9 OUT = NUM + DEN;

  // Update memory
  X.block<9, 1>(0, 2 * stage + 1) = X.block<9, 1>(0, 2 * stage);
  X.block<9, 1>(0, 2 * stage) = IN;
  Y.block<9, 1>(0, 2 * stage + 1) = Y.block<9, 1>(0, 2 * stage);
  Y.block<9, 1>(0, 2 * stage) = OUT;

  return OUT;
}

Vector9 DigitalNF(Vector9 IN, bool GND, float THRUST, float dT, Matrix9_4 &X, Matrix9_4 &Y) {
  Vector9 OUT;

  Matrix2_2 TRACK;
  TRACK << 1.3525, 42.6278, 2.6867, 84.8000;

  float fs = 1 / dT;
  float width1 = 35; // Hz

  // Sequential Notch Filter
  if (!GND) {
    // Notch #1
    // Setup second notch following track #1
    float f0 = TRACK(0, 0) * THRUST + TRACK(0, 1); // Hz
    IN = FilterStage(IN, f0, fs, width1, 0, X, Y);

    // 2nd Order Butterworth Filter
    // Set up a 2nd Order Butterworth Filter
    f0 = 20;
    float C = tan(M_PI * f0 / fs);
    float A1 = 2 * (C * C - 1) / (1 + sqrt(2) * C + C * C);
    float A2 = (1 - sqrt(2) * C + C * C) / (1 + sqrt(2) * C + C * C);
    float B0 = C * C / (1 + sqrt(2) * C + C * C);
    float B1 = 2 * B0;
    float B2 = B0;
    OUT = B0 * IN + B1 * X.block<9, 1>(0, 2) + B2 * X.block<9, 1>(0, 3) - A1 * Y.block<9, 1>(0, 2) - A2 * Y.block<9, 1>(0, 3);

    X.block<9, 1>(0, 3) = X.block<9, 1>(0, 2);
    X.block<9, 1>(0, 2) = IN;
    Y.block<9, 1>(0, 3) = Y.block<9, 1>(0, 2);
    Y.block<9, 1>(0, 2) = OUT;
  } else {
    // The output is just the input(passthrough)
    OUT = IN;

    // --- Update states sequentially for a smooth switch-on ---

    // Stage 1: History is updated with IN and OUT
    X.block<9, 1>(0, 1) = X.block<9, 1>(0, 0);
    X.block<9, 1>(0, 0) = IN;
    Y.block<9, 1>(0, 1) = Y.block<9, 1>(0, 0);
    Y.block<9, 1>(0, 0) = OUT;

    // The input to stage 2 is the output of stage 1
    Vector9 IN_2 = OUT;
    Vector9 OUT_2 = IN_2; // Passthrough for stage 2

    // Stage 2: History is updated
    X.block<9, 1>(0, 3) = X.block<9, 1>(0, 2);
    X.block<9, 1>(0, 2) = IN_2;
    Y.block<9, 1>(0, 3) = Y.block<9, 1>(0, 2);
    Y.block<9, 1>(0, 2) = OUT_2;

    // The final output is the passthrough from the last stage
    OUT = OUT_2;
  }

  return OUT;
}